docs/contributor-pin-signature-internals.rst

```diff
@@ -0,0 +1,780 @@
+
+IOModelOptions TypedDict
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Defines all options for configuring I/O pins:
+
+.. code-block:: python
+
+    class IOModelOptions(TypedDict):
+        invert: NotRequired[bool | Tuple[bool, ...]]
+        individual_oe: NotRequired[bool]
+        power_domain: NotRequired[str]
+        clock_domain: NotRequired[str]
+        buffer_in: NotRequired[bool]
+        buffer_out: NotRequired[bool]
+        sky130_drive_mode: NotRequired[Sky130DriveMode]
+        trip_point: NotRequired[IOTripPoint]
+        init: NotRequired[int | bool]
+        init_oe: NotRequired[int | bool]
+
+All fields use ``NotRequired`` to make them optional with sensible defaults.
+
+IOModel TypedDict
+~~~~~~~~~~~~~~~~~
+
+Extends ``IOModelOptions`` with direction and width information:
+
+.. code-block:: python
+
+    class IOModel(IOModelOptions):
+        direction: IODirection  # "input", "output", or "bidir"
+        width: int
+
+This is the complete model that gets annotated on I/O signatures.
+
+IOSignature Base Class
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The base class for all I/O signatures, decorated with ``@amaranth_annotate``:
+
+.. code-block:: python
+
+    @amaranth_annotate(IOModel, IO_ANNOTATION_SCHEMA, '_model')
+    class IOSignature(wiring.Signature):
+        def __init__(self, width: int, direction: IODirection, **kwargs: Unpack[IOModelOptions]):
+            # Build the model from parameters
+            model = IOModel(direction=direction, width=width, **kwargs)
+
+            # Create appropriate signal structure based on direction
+            if direction == "input":
+                members = {"i": In(width)}
+            elif direction == "output":
+                members = {
+                    "o": Out(width),
+                    "oe": Out(1) if not individual_oe else Out(width)
+                }
+            elif direction == "bidir":
+                members = {
+                    "i": In(width),
+                    "o": Out(width),
+                    "oe": Out(1) if not individual_oe else Out(width)
+                }
+
+            # Store model for annotation extraction
+            self._model = model
+
+            super().__init__(members)
+
+**Direction-Specific Subclasses:**
+
+.. code-block:: python
+
+    class InputIOSignature(IOSignature):
+        def __init__(self, width: int, **kwargs):
+            super().__init__(width, "input", **kwargs)
+
+    class OutputIOSignature(IOSignature):
+        def __init__(self, width: int, **kwargs):
+            super().__init__(width, "output", **kwargs)
+
+    class BidirIOSignature(IOSignature):
+        def __init__(self, width: int, **kwargs):
+            super().__init__(width, "bidir", **kwargs)
+
+Concrete Pin Signatures
+------------------------
+
+Core Module: ``chipflow_lib/platform/io/signatures.py``
+
+Concrete pin signatures (UART, GPIO, SPI, etc.) combine I/O signatures with simulation metadata.
+
```

Comment:
These are really annotitions of the *type* of the external interface, to allow for selection/typechecking of suitable simulation models that interface.

These are independant of the implmentation of the interface